sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -36,46 +36,91 @@ import org.apache.spark.util.collection.CompactBuffer
 case class SortMergeJoin(
     leftKeys: Seq[Expression],
     rightKeys: Seq[Expression],
+    joinType: JoinType,
     left: SparkPlan,
-    right: SparkPlan) extends BinaryNode {
+    right: SparkPlan,
+    condition: Option[Expression] = None) extends BinaryNode {
 
-  override def output: Seq[Attribute] = left.output ++ right.output
+  val (streamed, buffered, streamedKeys, bufferedKeys) = joinType match {
+    case RightOuter => (right, left, rightKeys, leftKeys)
+    case _ => (left, right, leftKeys, rightKeys)
+  }
+
+  override def output: Seq[Attribute] = joinType match {
+    case Inner =>
+      left.output ++ right.output
+    case LeftOuter =>
+      left.output ++ right.output.map(_.withNullability(true))
+    case RightOuter =>
+      left.output.map(_.withNullability(true)) ++ right.output
+    case FullOuter =>
+      left.output.map(_.withNullability(true)) ++ right.output.map(_.withNullability(true))
+    case x =>
+      throw new IllegalStateException(s"SortMergeJoin should not take $x as the JoinType")
+  }
 
-  override def outputPartitioning: Partitioning = left.outputPartitioning
+  override def outputPartitioning: Partitioning = joinType match {
+    case FullOuter =>
+      // when doing Full Outer join, NULL rows from both sides are not so partitioned.
+      UnknownPartitioning(streamed.outputPartitioning.numPartitions)
+    case _ => streamed.outputPartitioning
```

Comment:
Both `streamed` and `buffered` have the same output partitioning, since we've co-partitioned them in order to perform the join? Just want to double-check my understanding.